contracts/pcv/morpho/MorphoCompoundPCVDeposit.sol

```diff
@@ -66,54 +129,256 @@ contract MorphoCompoundPCVDeposit is PCVDeposit {
+        cTokens[0] = cToken;
+
+        /// set swap comp to morpho flag false to claim comp rewards
+        uint256 claimedAmount = IMorpho(morpho).claimRewards(cTokens, false);
+
+        emit Harvest(COMP, int256(claimedAmount), block.timestamp);
+    }
+
+    /// @notice function that emits an event tracking profits and losses
+    /// since the last contract interaction
+    /// then writes the current amount of PCV tracked in this contract
+    /// to lastRecordedBalance
+    /// @return the amount deposited after adding accrued interest or realizing losses
+    function accrue() external nonReentrant whenNotPaused returns (uint256) {
+        int256 startingRecordedBalance = lastRecordedBalance.toInt256();
+
+        _recordPNL(); /// update deposit amount and fire harvest event
+
+        int256 endingRecordedBalance = lastRecordedBalance.toInt256();
+
+        if (pcvOracle != address(0)) {
+            /// if any amount of PCV is withdrawn and no gains, delta is negative
+            IPCVOracle(pcvOracle).updateLiquidBalance(
+                endingRecordedBalance - startingRecordedBalance
+            );
+        }
+
+        return lastRecordedBalance; /// return updated pcv amount
+    }
+
+    /// ------------------------------------------
+    /// ------------ Permissioned API ------------
+    /// ------------------------------------------
+
     /// @notice withdraw tokens from the PCV allocation
+    /// non-reentrant as state changes and external calls are made
     /// @param to the address PCV will be sent to
     /// @param amount of tokens withdrawn
-    function withdraw(address to, uint256 amount) external onlyPCVController {
-        _withdraw(to, amount);
+    function withdraw(address to, uint256 amount)
+        external
+        onlyPCVController
+        nonReentrant
+    {
+        int256 startingRecordedBalance = lastRecordedBalance.toInt256();
+
+        _withdraw(to, amount, true);
+
+        int256 endingRecordedBalance = lastRecordedBalance.toInt256();
+        if (pcvOracle != address(0)) {
+            /// if any amount of PCV is withdrawn and no gains, delta is negative
+            IPCVOracle(pcvOracle).updateLiquidBalance(
+                endingRecordedBalance - startingRecordedBalance
+            );
+        }
     }
 
     /// @notice withdraw all tokens from Morpho
+    /// non-reentrant as state changes and external calls are made
     /// @param to the address PCV will be sent to
-    function withdrawAll(address to) external onlyPCVController {
-        uint256 amount = balance();
-        _withdraw(to, amount);
+    function withdrawAll(address to) external onlyPCVController nonReentrant {
+        int256 startingRecordedBalance = lastRecordedBalance.toInt256();
+
+        /// compute profit from interest accrued and emit an event
+        _recordPNL();
+
+        /// withdraw last recorded amount as this was updated in record pnl
+        _withdraw(to, lastRecordedBalance, false);
```

Comment:
For a withdraw all, you also can pass `type(uint256).max` as argument to avoid leaving dust on Morpho but as you're calling `_recordPNL` just before it must be exactly the same.

Reply:
Passing uint256 max to the withdraw function would revert as lastRecordedBalance has amount subtracted from it. This would cause an underflow and revert.